b2sdk/replication/monitoring.py

```diff
@@ -0,0 +1,281 @@
+
+    @property
+    def source_folder(self) -> B2_FOLDER_CLASS:
+        return self.B2_FOLDER_CLASS(
+            bucket_name=self.bucket.name,
+            folder_name=self.rule.file_name_prefix,
+            api=self.source_api,
+        )
+
+    @property
+    def destination_bucket(self) -> Bucket:
+        destination_api = self.destination_api or self.source_api
+        bucket_id = self.rule.destination_bucket_id
+        return destination_api.get_bucket_by_id(bucket_id)
+
+    @property
+    def destination_folder(self) -> B2_FOLDER_CLASS:
+        destination_bucket = self.destination_bucket
+        return self.B2_FOLDER_CLASS(
+            bucket_name=destination_bucket.name,
+            folder_name=self.rule.file_name_prefix,
+            api=destination_bucket.api,
+        )
+
+    def iter_pairs(self) -> Iterator[Tuple[Optional[B2Path], Optional[B2Path]]]:
+        """
+        Iterate over files in source and destination and yield pairs.
+        Required for replication inspection in-depth.
+
+        Return pair of (source B2Path, destination B2Path). Source or destination
+        path may be missing if there's no corresponding destination/source file.
+        """
+        yield from zip_folders(
+            self.source_folder,
+            self.destination_folder,
+            reporter=self.report,
+            policies_manager=self.scan_policies_manager,
+        )
+
+    def scan(self, scan_destination: bool = True) -> AbstractReplicationReport:
+        """
+        Scan source bucket (only, or with destination) and return replication report.
+
+        No destination scan may give limited replication information, since it only
+        checks files on the source bucket without checking whether
+        they we really replicated to destination. It may be handy though
+        if there is no access to replication destination.
+        """
+        report = self.replication_report_class()
+
+        thread_pool = ThreadPoolExecutor(max_workers=1)
+        queue = Queue(maxsize=self.QUEUE_SIZE)
+
+        with ThreadPoolExecutor(max_workers=2) as thread_pool:
+            if not scan_destination:
+
+                def fill_queue():
+                    for path in self.source_folder.all_files(
+                        policies_manager=self.scan_policies_manager,
+                        reporter=self.report,
+                    ):
+                        queue.put(path, block=True)
+
+                def consume_queue():
+                    while not queue.empty():
+                        path = queue.get(block=True)
+                        report.add(path)
+
+            else:
+
+                def fill_queue():
+                    for pair in self.iter_pairs():
+                        queue.put(pair, block=True)
+
+                def consume_queue():
+                    while not queue.empty():
+                        pair = queue.get(block=True)
+                        report.add(*pair)
```

Comment:
I think this can have a common implementation for both `if` branches
```suggestion
            def consume_queue():
                while not queue.empty():
                    items = queue.get(block=True)
                    report.add(*items)
```
and it'll work fine for 1-element lists